src/meta/src/hummock/compaction/compaction_picker.rs

```diff
@@ -39,15 +40,66 @@ impl MinOverlappingPicker {
         compact_task_id: u64,
         level: usize,
         overlap_strategy: Arc<dyn OverlapStrategy>,
+        config: Arc<CompactionConfig>,
     ) -> MinOverlappingPicker {
         MinOverlappingPicker {
             compact_task_id,
             overlap_strategy,
+            config,
             level,
         }
     }
 }
 
+impl MinOverlappingPicker {
+    // For example:
+    //  L1:      [(k0, k4),  (k5, k7), (k8, k9)]
+    //  L2:      [(k0, k1), (k2, k6)]
+    // If we only choose file (k0, k4), (k0, k1), (k2, k6), the file (k5, k7) will compact with the
+    // result of their compaction task again. So a better strategy is choosing (k0, k4), (k0, k1),
+    // (k2, k6), (k5, k7) all in one task.
+    fn try_expand_input(
+        &self,
+        levels: &[Level],
+        level_handlers: &mut [LevelHandler],
+        select_input_ssts: Vec<SstableInfo>,
+        target_input_ssts: Vec<SstableInfo>,
+    ) -> (Vec<SstableInfo>, Vec<SstableInfo>) {
+        let target_level = self.level + 1;
+        let select_overlap_files = self
+            .overlap_strategy
+            .check_base_level_overlap(&target_input_ssts, &levels[self.level].table_infos);
+        if select_overlap_files
+            .iter()
+            .any(|table| level_handlers[self.level].is_pending_compact(&table.id))
+        {
+            return (select_input_ssts, target_input_ssts);
+        }
+        let target_overlap_files = self
+            .overlap_strategy
+            .check_base_level_overlap(&select_overlap_files, &levels[target_level].table_infos);
+        if target_overlap_files
```

Comment:
LevelDB only expands select level SSTs without expanding the target level SSTs. We expand both here. I wonder whether this will cause the same situation mentioned in L55-60 since after expanding target level SSTs, the key range may also expand which overlaps with more unselected select level SSTs.

Reply:
No. If expand select level without expanding the target level SSTS, this expand may be wrong....

Reply:
Because if we do not expand the target level SSTs again, this expanded select files may overlap with some files which has not been picked up

Reply:
I understand. LevelDB handles this problem by only trying to expand SSTs in select level without growing targe level. 

For example:
```
L1:      [(k0, k4),  (k5, k7), (k8, k12)]
L2:      [(k0, k9), (k10, k15)]
```

We start with (k0, k4) in L1 and (k0, k9) in L2. During expansion, we only add (k5, k7) in L1 but not (k8, k12) becuase adding (k8, k12) requires expanding L2.

Reply:
Ok, I get it.....